### PR TITLE
Use np.random's RandomState when seed is None

### LINKED ISSUE #13143: sample not using numpy's random state
After fixing a random seed with ``numpy.random.seed``, I expect ``sample`` to yield the same results.

Expected behavior of ``numpy.random.choice`` but found something different. Here is pandas:

```python
In [1]: import pandas as pd

In [2]: import numpy as np

In [3]: df = pd.DataFrame(np.arange(1000))
In [12]: np.random.seed(5); df.sample(2)
Out[12]: 
       0
824  824
225  225

In [13]: np.random.seed(5); df.sample(2)
Out[13]: 
       0
182  182
586  586
```

Whereas numpy.random.choice is consistent

```
In [6]: np.random.seed(5); np.random.choice(1000)
Out[6]: 867

In [7]: np.random.seed(5); np.random.choice(1000)
Out[7]: 867
```

#### output of ``pd.show_versions()``

```
In [8]: pd.show_versions()

INSTALLED VERSIONS
------------------
commit: None
python: 3.4.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.16.0-67-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.1
nose: 1.3.1
pip: 8.1.1
setuptools: 18.4
Cython: 0.23.4
numpy: 1.11.0
scipy: 0.16.1
statsmodels: 0.6.1
xarray: None
IPython: 4.0.1
sphinx: 1.3.1
patsy: 0.4.0
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: None
tables: 3.2.2
numexpr: 2.4.4
matplotlib: 1.5.0
openpyxl: None
xlrd: 0.9.4
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.2.1
html5lib: 0.999
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: None
```

### PR DESCRIPTION
**Pull request by ariddell**
- [x] closes #13143
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

The handle for numpy's current random state is
`np.random.mtrand._rand`.

Compare https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/validation.py#L573

### CODE REVIEW

File: pandas/core/common.py [hunk 1]
```diff
@@ -2072,7 +2072,7 @@ def _random_state(state=None):
     elif isinstance(state, np.random.RandomState):
         return state
     elif state is None:
-        return np.random.RandomState()
+        return np.random.mtrand._rand
```
Comment (shoyer) [inline pandas/core/common.py]: Can we use something like this instead?
```python
seed = np.random.randint(2 * 32)
return np.random.RandomState(seed)
```

That way we don't need to use NumPy's private API.

File: pandas/tests/test_generic.py [hunk 2]
```diff
@@ -415,6 +415,10 @@ def test_sample(self):
                 o.sample(frac=0.7, random_state=np.random.RandomState(test)),
                 o.sample(frac=0.7, random_state=np.random.RandomState(test)))
 
+            np.random.seed(test)
+            self._compare(o.sample(n=4), o.sample(n=4))
```
Comment (shoyer) [inline pandas/tests/test_generic.py]: This test doesn't look right to me. You need to call `np.random.seed(test)` before every call to `sample`, not just the first one.

File: pandas/core/common.py (hunk 1 above)
Comment (jorisvandenbossche) [inline pandas/core/common.py]: I don't think that will do the same?

As this is to catch a previously called `np.random.seed()`, so feeding it another seed will not give the desired result I think?

Comment (shoyer) [inline pandas/core/common.py]: We want results to be reproducible after calling `np.random.seed()`. But we don't need to reuse exactly the same seed -- it's OK to also use a seed derived from NumPy's seed.

Using the exact same seed would only be necessary if we want to promise the `sample` makes the exact same choice as `np.random.choice`. But that's really an implementation detail.

(Clearly a comment in the code would also be in order if we use my suggested approach.)

Comment (jorisvandenbossche) [inline pandas/core/common.py]: Aha, I see. 
For me it does not really matter which approach we take (sklearn also uses `_rand`, so although it is private it seems ok)

File: pandas/tests/test_generic.py (hunk 2 above)
Comment (ariddell) [inline pandas/tests/test_generic.py]: You're right. I fixed the test.

**Comment by jreback** [comment]
@ariddell pls also add a whatsnew entry. I would do it in API changes section.

**Comment by ariddell** [comment]
@jreback added. Thanks!

**Comment by shoyer** [comment]
@ariddell could you please fix the use of NumPy's private state as noted above? This is something simple that could avoid significant pain down the road.

**Comment by ariddell** [comment]
Fair enough. What should the seed be? Did you mean ``2 ** 32`` above?

**Comment by shoyer** [comment]
@ariddell Yes, I meant `2 ** 32` :).

**Comment by rkern** [comment]
Using `np.random.mtrand._rand` is the right approach. It's not going to disappear. This is an authorized use.

**Comment by shoyer** [comment]
Let's just return `np.random` from `_random_state` instead of `np.random.mtrand._rand`, using @rkern's suggestion of duck typing for RandomState objects from the related thread on numpy-discussion: https://mail.scipy.org/pipermail/numpy-discussion/2016-May/075489.html

**Comment by ariddell** [comment]
@shoyer duck typing it is.

File: pandas/core/common.py [hunk 3]
```diff
@@ -2072,7 +2072,7 @@ def _random_state(state=None):
     elif isinstance(state, np.random.RandomState):
         return state
     elif state is None:
-        return np.random.RandomState()
+        return np.random
```
Comment (shoyer) [inline pandas/core/common.py]: can you please update the docstring, too?

Comment (ariddell) [inline pandas/core/common.py]: Yes. Fixed.

**Comment by jreback** [comment]
you have a failing test

**Comment by jreback** [comment]
@ariddell can you rebase / update

**Comment by ariddell** [comment]
OK. I think I've got it.

**Comment by jreback** [comment]
thanks @ariddell

**Comment by pydolan** [comment]
I'm not getting the behavior I'd expect from this fix in Pandas 0.19.2. Example:

```
>>> import pandas as pd
>>> import numpy as np
>>> df = pd.DataFrame(np.random.randn(10, 5), columns=['a', 'b', 'c', 'd', 'e'])
>>> df.sample(n=2)
          a         b         c         d         e
8  0.365786 -0.855795 -0.511062 -0.116734  1.133541
0  1.281455 -0.423852  0.011236 -0.945194 -0.088124
>>> df.sample(n=2)
          a         b         c         d         e
2  0.341751 -1.158737  0.313814  1.827552 -0.351045
9  1.355003  0.532651  0.250463 -0.281751 -0.342741
>>> df.sample(n=2, random_state=12345678)
          a         b         c         d         e
8  0.365786 -0.855795 -0.511062 -0.116734  1.133541
6 -0.401077 -0.047115 -0.410951 -1.608354  0.290594
>>> df.sample(n=2, random_state=12345678)
          a         b         c         d         e
8  0.365786 -0.855795 -0.511062 -0.116734  1.133541
6 -0.401077 -0.047115 -0.410951 -1.608354  0.290594
>>> pd.__version__
u'0.19.2'
>>> np.__version__
'1.12.0'
```

Any thoughts on why `df.sample(n=2)` doesn't match `df.sample(n=2, random_state=12345678)`?

**Comment by shoyer** [comment]
@pydolan is there a missing `np.random.seed()` call somewhere in your example?

**Comment by pydolan** [comment]
@shoyer -- sorry about that; will update example with my actual code (i.e., with the one missing line)

**Comment by shoyer** [comment]
@pydolan It looks like you are calling `np.random.randn()` after setting the seed before `df.sample()`

**Comment by pydolan** [comment]
If I move the `seed()` call after instantiating the dataframe, I still get inconsistent behavior with calls to `sample()`, except when I provide the `random_state` arg. Example:

```
>>> import pandas as pd
>>> import numpy as np
>>> df = pd.DataFrame(np.random.randn(10, 5), columns=['a', 'b', 'c', 'd', 'e'])
>>> np.random.seed(12345678)
>>> df.sample(n=2)
          a         b         c         d         e
8 -1.250204  0.551508  1.408080  0.397452  0.424326
6 -0.028298  0.203270  0.939094 -1.802227 -0.088679
>>> df.sample(n=2)
          a         b         c         d         e
4  0.895497  0.609853 -1.548664 -1.238415 -1.058904
5  0.196420  0.472877 -0.918205  1.019862 -0.631993
>>> df.sample(n=2, random_state=12345678)
          a         b         c         d         e
8 -1.250204  0.551508  1.408080  0.397452  0.424326
6 -0.028298  0.203270  0.939094 -1.802227 -0.088679
>>> df.sample(n=2, random_state=12345678)
          a         b         c         d         e
8 -1.250204  0.551508  1.408080  0.397452  0.424326
6 -0.028298  0.203270  0.939094 -1.802227 -0.088679
```

Note that this time, the first call to `sample()` uses the seed, but the second call does not use the seed. Is it expected that `seed()` needs to be called before every sample call? I thought it is supposed to be, "set once", and all future randomization-related calls should use (including my original example, where `randn()` is called after `seed()` and before `sample()`).

(Also note that I did verify that calling `seed()` before every call to `sample()` does indeed produce the same sampled rows.)

**Comment by shoyer** [comment]
> Is it expected that seed() needs to be called before every sample call? I thought it is supposed to be, "set once", and all future randomization-related calls should use (including my original example, where randn() is called after seed() and before sample()).

This is working as intended. Try just sampling random numbers from numpy -- it does the exact same thing.

**Comment by pydolan** [comment]
Good point; I completely spaced on the nature of random sampling. Sorry for the waste of time!
